File: MPI_Code/Fortran/subarray.f90
PROGRAM subarray
USE mpi_f08
IMPLICIT NONE
INTEGER,PARAMETER::ndims=2
INTEGER::myrank,i,j
INTEGER::ibuf(2:7, 0:6)
INTEGER::arr_sizes(ndims), arr_subsizes(ndims), arr_starts(ndims)
TYPE(MPI_Datatype)::newtype
CALL MPI_Init
CALL MPI_Comm_rank(MPI_COMM_WORLD,myrank)
ibuf=0
IF(myrank==0)THEN
  DO j=0,6
    DO i=2,7
      ibuf(i,j)=i    
    END DO
  END DO
ENDIF
arr_sizes=(/6,7/);   arr_subsizes=(/2,5/);  arr_starts=(/1,1/)
CALL MPI_Type_create_subarray(ndims,arr_sizes,arr_subsizes,arr_starts,MPI_ORDER_FORTRAN,MPI_INTEGER,newtype)
CALL MPI_Type_commit(newtype)
CALL MPI_Bcast(ibuf,1,newtype,0,MPI_COMM_WORLD)
IF(MYRANK==1)THEN
  DO i=2,7
    PRINT*,ibuf(i,:)
  END DO
ENDIF
CALL MPI_Type_free(newtype)
CALL MPI_Finalize
END PROGRAM subarray

File: example_code/mpi/MPI_Code/C/subarray.c
#include <stdio.h>
#include "mpi.h"
#define ndims 2
int main()
{
  int ibuf[6][7];
  int arr_sizes[ndims]={6,7}, arr_subsizes[ndims]={2,5}, arr_starts[ndims]={1,1};
  int i,j, myrank;
  MPI_Datatype newtype;
  MPI_Init(NULL,NULL);
  MPI_Comm_rank(MPI_COMM_WORLD,&myrank);
  if(myrank==0) {
    for(i=0;i<6;i++)
      for(j=0;j<7;j++) ibuf[i][j]=i+2;
  }else{
    for(i=0;i<6;i++)
      for(j=0;j<7;j++) ibuf[i][j]=0;
  }
  MPI_Type_create_subarray(ndims,arr_sizes,arr_subsizes,arr_starts,MPI_ORDER_C, \
                           MPI_INT,&newtype);
  MPI_Type_commit(&newtype);
  MPI_Bcast(ibuf,1,newtype,0,MPI_COMM_WORLD);
  if(myrank==1){
    for(i=0;i<6;i++){
      for(j=0;j<7;j++) printf(" %d",ibuf[i][j]);
      printf("\n");
   }
  }
  MPI_Type_free(&newtype);
  MPI_Finalize();
  return 0;
}
